mmseg/models/losses/hausdorff_distance_loss.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+"""Modified from
+https://github.com/JunMa11/SegWithDistMap/blob/master/code/train_LA_HD.py#L106
+(Apache-2.0 license)"""
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from scipy.ndimage import distance_transform_edt
+
+from ..builder import LOSSES
+from .utils import get_class_weight, weighted_loss
+
+
+def compute_dtm(mask_arr, normalized=False):
+    """compute the distance transform map of foreground in binary mask.
+
+    Args:
+        mask_arr (array): segmentation mask, shape=(batch_size, x, y).
+                          or shape=(batch_size, x, y, z)
+        normalized (bool): whether compute the dtm by normalized.
+                           Default to False.
+
+    Returns:
+        dtm (array): the foreground Distance Map (SDM), shape=out_shape.
+            dtm(x) = 0;    x out of segmentation or x in segmentation boundary.
+                   = inf|x - y|;    x in segmentation.
+                                    if normalized is True, dtm(x) to [0, 1].
+    """
+    dtm = np.zeros_like(mask_arr)
+
+    for batch in range(len(mask_arr)):
+        fg_mask = mask_arr[batch] > 0.5
+        if fg_mask.any():
+            bg_mask = ~fg_mask
+            fg_distance = distance_transform_edt(fg_mask)
+            bg_distance = distance_transform_edt(bg_mask)
+            if normalized:
+                fg_distance = fg_distance / np.max(fg_distance)
+                bg_distance = bg_distance / np.max(bg_distance)
+            dtm[batch] = fg_distance + bg_distance
+    return dtm
+
+
+@weighted_loss
+def binary_hausdorff_distance_loss(pred, target, valid_mask):
+    assert pred.shape[0] == target.shape[0]
+
+    with torch.no_grad():
+        gt_dtm_npy = compute_dtm((target * valid_mask).numpy())
```

Comment:
This line may raise error when using GPU because it needs `.cpu()` operation. Otherwise `TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.`

Reply:
ok